lib/util.js

```diff
@@ -72,6 +71,79 @@ const errorToString = Error.prototype.toString;
 var CIRCULAR_ERROR_MESSAGE;
 var Debug;
 
+// eslint-disable-next-line
+const escapeSequencesRegExp = /[\\'\x00-\x1f]/g;
+const keyStrRegExp = /^[a-zA-Z_][a-zA-Z_0-9]*$/;
+const textRegExp = /^[a-zA-Z_0-9 ,.?!+\-&"öäüÜÖÄß():;<>{}[\]=$]*$/;
+const colorRegExp = /\u001b\[\d\d?m/g;
+const numberRegExp = /^(0|[1-9][0-9]*)$/;
+
+// Escaped special characters. Use empty strings to fill up unused entries.
+const meta = [
+  '\\u0000', '\\u0001', '\\u0002', '\\u0003', '\\u0004',
+  '\\u0005', '\\u0006', '\\u0007', '\\b', '\\t',
+  '\\n', '\\u000b', '\\f', '\\r', '\\u000e',
+  '\\u000f', '\\u0010', '\\u0011', '\\u0012', '\\u0013',
+  '\\u0014', '\\u0015', '\\u0016', '\\u0017', '\\u0018',
+  '\\u0019', '\\u001a', '\\u001b', '\\u001c', '\\u001d',
+  '\\u001e', '\\u001f', '', '', '',
+  '', '', '', '', "\\'", '', '', '', '', '',
+  '', '', '', '', '', '', '', '', '', '',
+  '', '', '', '', '', '', '', '', '', '',
+  '', '', '', '', '', '', '', '', '', '',
+  '', '', '', '', '', '', '', '', '', '',
+  '', '', '', '', '', '', '', '\\\\'
+];
+
+// Escape control characters, single quotes and the backslash.
+// This is similar to JSON stringify escaping.
+function strEscape(str) {
+  // Some magic numbers that worked out fine while benchmarking with v8 6.0
+  if (str.length < 5000 && textRegExp.test(str))
+    return `'${str}'`;
+  if (str.length > 50) {
+    return `'${str.replace(escapeSequencesRegExp, (str) => {
+      return meta[str.charCodeAt(0)];
+    })}'`;
+  }
+  var result = '';
+  var last = 0;
+  for (var i = 0; i < str.length; i++) {
+    const point = str.charCodeAt(i);
+    if (point === 39 || point === 92 || point < 32) {
+      result += `${str.slice(last, i)}${meta[point]}`;
```

Comment:
You might consider adding a check that `last < i` to avoid unnecessarily slicing to an empty string. That's typically what I've done in other parts of core.

Reply:
I am not a big fan of adding this extra check as it makes the code more complicated for rare cases but I will add the check nevertheless.

Reply:
I wouldn't think they would really be that rare. Consecutive characters like tabs or newlines for example could be more common. Adding the extra check may/may not be worth it though, it would need some benchmarking to confirm my suggestion improves things.

Reply:
Changed.